### PR TITLE
[Example] fix cpp example inception-bn and training acc issue

### PR DESCRIPTION
**Pull request by roywei**
## Description ##
1. Fix: #9417
Root Cause:
Passing correct padding (1, 1) in Pooling in InceptionFactoryB
Reference python implementation, `pad=(1, 1)`
Tested example working fine.

2. fix some models not training: https://github.com/apache/incubator-mxnet/issues/13243, https://github.com/apache/incubator-mxnet/issues/12966, https://github.com/apache/incubator-mxnet/issues/8108
Root Cause:
Wrong logic for gradient requirement: symbols with name length 3 and 4 won't have gradients
Changed to:
Symbols names contains "data" or "label" should not be required for gradients

3. Added parameter initialization same as python versions

4. added inception-bn to unit tests

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
see description

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: cpp-package/example/inception_bn.cpp [hunk 1]
```diff
@@ -172,7 +173,13 @@ int main(int argc, char const *argv[]) {
   auto val_iter = MXDataIter("MNISTIter");
   setDataIter(&val_iter, "Label", data_files, batch_size);
 
-  Optimizer* opt = OptimizerRegistry::Find("ccsgd");
```
Comment (stu1130) [inline cpp-package/example/inception_bn.cpp]: is there any reason to change the optimizer?

File: cpp-package/tests/ci_test.sh [hunk 2]
```diff
@@ -36,6 +36,9 @@ cp ../../build/cpp-package/example/lenet_with_mxdataiter .
 cp ../../build/cpp-package/example/resnet .
 ./resnet 5
 
+cp ../../build/cpp-package/example/resnet .
+./inception-bn 5
+
 cp ../../build/cpp-package/example/mlp .
```
Comment (stu1130) [inline cpp-package/tests/ci_test.sh]: should we also add the mlp_csv example here?

Comment (roywei) [reply cpp-package/tests/ci_test.sh]: @stu1130 `ccsgd` was deprecated long time ago, actually, I will update to change `ccsgd` in all other examples.

**Comment by kalyc** [comment]
@roywei please take a look at the failed CI and re-trigger it

**Comment by stu1130** [comment]
@mxnet-label-bot add [pr-awaiting-review]

**Comment by roywei** [comment]
@marcoabreu according to dev list discussion, this PR will be blocked by https://github.com/apache/incubator-mxnet/pull/13344 as it's changing Jenkins file

File: Jenkinsfile [hunk 3]
```diff
@@ -35,7 +35,7 @@ mx_cmake_lib_debug = 'build/libmxnet.so, build/libmxnet.a, build/3rdparty/dmlc-c
...
-mx_lib_cpp_examples = 'lib/libmxnet.so, lib/libmxnet.a, 3rdparty/dmlc-core/libdmlc.a, 3rdparty/tvm/nnvm/lib/libnnvm.a, 3rdparty/ps-lite/build/libps.a, deps/lib/libprotobuf-lite.a, deps/lib/libzmq.a, build/cpp-package/example/lenet, build/cpp-package/example/alexnet, build/cpp-package/example/googlenet, build/cpp-package/example/lenet_with_mxdataiter, build/cpp-package/example/resnet, build/cpp-package/example/mlp, build/cpp-package/example/mlp_cpu, build/cpp-package/example/mlp_gpu, build/cpp-package/example/test_score, build/cpp-package/example/test_optimizer'
+mx_lib_cpp_examples = 'lib/libmxnet.so, lib/libmxnet.a, 3rdparty/dmlc-core/libdmlc.a, 3rdparty/tvm/nnvm/lib/libnnvm.a, 3rdparty/ps-lite/build/libps.a, deps/lib/libprotobuf-lite.a, deps/lib/libzmq.a, build/cpp-package/example/lenet, build/cpp-package/example/alexnet, build/cpp-package/example/googlenet, build/cpp-package/example/lenet_with_mxdataiter, build/cpp-package/example/resnet, build/cpp-package/example/inception_bn, build/cpp-package/example/mlp, build/cpp-package/example/mlp_cpu, build/cpp-package/example/mlp_gpu, build/cpp-package/example/test_score, build/cpp-package/example/test_optimizer'
```
Comment (nswamy) [inline Jenkinsfile]: Please move jenkinsFile update to a different PR

**Comment by roywei** [comment]
@nswamy I created separate PR https://github.com/apache/incubator-mxnet/pull/13367, lets merge this if CI passed.
Thanks!
